python/ray/util/sgd/data/dataset.py

```diff
@@ -0,0 +1,127 @@
+from abc import ABC, abstractmethod
+
+import random
+import ray
+import torch
+
+from ray.util.sgd.data.utils import get_arr_partition
+
+
+@ray.remote
+def _load_remote(dataset, path):
+    return dataset.load(path)
+
+
+class Dataset(ABC, torch.utils.data.IterableDataset):
+    """
+    This class defines the Ray dataset abstration (primarily for SGD).
+    It's primarily meant to be treated as an abstract class.
+    """
+
+    def __init__(self, loc, remote=False, max_paths=None, transform=None):
+        """Needs to be initialized with a dataset location
+
+        Args:
+            loc: Location of training data (i.e. root of training directory)
+                This is passed into get_paths
+            remote: Whether dataset loading
+                should be parallelized on multiple workers/nodes. For example,
+                if only the driver has the necessary permissions to access
+                the data, remote should be False
+            max_paths: Maximum number of paths to load (passed to `get_paths`)
+
+        """
+        self._paths = self.get_paths(loc, max_paths=max_paths)
+        self.transform = transform
+
+        if remote:
+            data = [_load_remote.remote(self, path) for path in self._paths]
```

Comment:
This remote might end up on a different node (AWS instance) and download the data to that machine. It will then be serialized, put into plasma, and sent over to the local node, effectively re-downloading the data. Not sure if this behavior is intended, but it seems really weird to me. Perhaps we should always force the `_load_remote` onto the current node?

Reply:
This is intentional, there are use cases (like s3) where due to throttling our bandwidth to the datastore is a function of the number of nodes downloading, so we want more nodes to download data and reshuffling is cheap because we have much higher internal bandwidth.

Reply:
I am interested to see more examples of this, but it makes more sense now, thanks!